ckb-importer/src/main/java/com/hartwig/serve/ckb/datamodel/treatmentapproaches/RelevantTreatmentApproachesFactory.java

```diff
@@ -19,49 +19,58 @@ private RelevantTreatmentApproachesFactory() {
+        List<DrugClassTreatmentApproach> relevantTreatmentApproach = Lists.newArrayList();
 
         for (TreatmentApproachInfo treatmentApproachInfo : treatmentApproachInfos) {
-            RelevantTreatmentApproaches resolvedRelevantTreatmentApproaches =
-                    resolveRelevantTreatmentApproaches(ckbJsonDatabase, treatmentApproachInfo);
+            DrugClassTreatmentApproach resolvedRelevantTreatmentApproaches =
+                    resolveDrugTreatmentApproaches(ckbJsonDatabase, treatmentApproachInfo);
             relevantTreatmentApproach.add(resolvedRelevantTreatmentApproaches);
         }
         return relevantTreatmentApproach;
     }
 
     @NotNull
-    private static TreatmentApproachIntervation extractTreatmentApproachIntervation(@NotNull CkbJsonDatabase ckbJsonDatabase,
-            @NotNull JsonTreatmentApproach treatmentApproach) {
-        boolean isDrugClass = treatmentApproach.drugClass() != null;
-        boolean isTherapy = treatmentApproach.therapy() != null;
+    public static List<TherapyTreatmentApproach> extractTherapyTreatmentApproaches(@NotNull CkbJsonDatabase ckbJsonDatabase,
+            @NotNull List<TreatmentApproachInfo> treatmentApproachInfos) {
+        List<TherapyTreatmentApproach> relevantTreatmentApproach = Lists.newArrayList();
 
-        if (isDrugClass && isTherapy) {
-            throw new IllegalStateException("An treatment approach cannot be both a drug class and therapy");
+        for (TreatmentApproachInfo treatmentApproachInfo : treatmentApproachInfos) {
+            TherapyTreatmentApproach resolvedRelevantTreatmentApproaches =
+                    resolveTherapyTreatmentApproaches(ckbJsonDatabase, treatmentApproachInfo);
+            relevantTreatmentApproach.add(resolvedRelevantTreatmentApproaches);
         }
+        return relevantTreatmentApproach;
+    }
 
-        if (isDrugClass) {
-            return ImmutableTherapyTreatmentApproach.builder()
-                    .therapy(TherapyFactory.resolveTherapy(ckbJsonDatabase, Objects.requireNonNull(treatmentApproach.therapy())))
-                    .build();
-        } else if (isTherapy) {
-            return ImmutableDrugClassTreatmentApproach.builder()
-                    .drugClass(DrugFactory.resolveDrugClass(ckbJsonDatabase, Objects.requireNonNull(treatmentApproach.drugClass())))
-                    .build();
-        } else {
-            throw new IllegalStateException("An treatment approach cannot be both a drug class and therapy!");
+    @NotNull
+    private static DrugClassTreatmentApproach resolveDrugTreatmentApproaches(@NotNull CkbJsonDatabase ckbJsonDatabase,
+            @NotNull TreatmentApproachInfo treatmentApproachInfo) {
+
+        for (JsonTreatmentApproach treatmentApproach : ckbJsonDatabase.treatmentApproaches()) {
+            if (treatmentApproach.id() == treatmentApproachInfo.id() && treatmentApproach.drugClass() != null) {
```

Comment:
This is a common antipattern in SERVE where we do a full DB scan for each ID we need to resolve instead of building a map first. I fixed this in `VariantFactory` in https://github.com/hartwigmedical/serve/pull/31 by creating an instance of the factory that holds the maps, but the issue is so prevalent we may be better off converting the lists in the `CkbJsonDatabase` to maps so we can support fast lookups with static use of the factories. This is out of scope here, but just curious if anyone else has thoughts. Also, I recognize that these changes are unlikely to significantly impact SERVE's runtime for most use cases 🙂

Reply:
I will leave this to another round or improvements